packages/internals/src/engine-commands/getEnginesMetaInfo.ts

```diff
@@ -0,0 +1,178 @@
+  }[] = engineDataAcc.map((arr) => arr[0])
+
+  // keep track of any error that has occurred, if any
+  const enginesMetaInfoErrors: Error[] = engineDataAcc.flatMap((arr) => arr[1])
+
+  return [engineMetaInfo, enginesMetaInfoErrors] as const
+}
+
+export function getEnginesInfo(enginesInfo: EngineInfo): readonly [string, Error[]] {
+  // if the engine is not found, or the version cannot be retrieved, keep track of the resulting errors.
+  const errors = [] as Error[]
+
+  // compute message displayed when an engine is resolved via env vars
+  const resolved = match(enginesInfo)
+    .with({ fromEnvVar: P.when(O.isSome) }, (enginesInfoLibrary) => {
+      return `, resolved by ${enginesInfoLibrary.fromEnvVar.value}`
+    })
+    .otherwise(() => '')
+
+  // compute version (git hash) of an engine, returning an error message and populating
+  // `errors` if it fails
+  const version = match(enginesInfo)
+    .with({ version: P.when(E.isRight) }, (engineInfo) => {
+      return engineInfo.version.right
+    })
+    .with({ version: P.when(E.isLeft) }, (engineInfo) => {
+      errors.push(engineInfo.version.left)
+      return 'E_CANNOT_RESOLVE_VERSION' as const
+    })
+    .otherwise(() => {
+      return 'E_CANNOT_RESOLVE_VERSION' as const
+    })
+
+  // compute absolute path of an engine, returning an error message and populating
+  // `errors` if it fails
+  const absolutePath = match(enginesInfo)
+    .with({ libraryPath: P.when(O.isSome) }, (enginesInfoLibrary) => {
+      return enginesInfoLibrary.libraryPath.value
+    })
+    .with({ libraryPath: P.when(O.isNone) }, (_) => {
+      return 'E_CANNOT_RESOLVE_PATH' as const
+    })
+    .with({ binaryPath: P.when(E.isRight) }, (engineInfo) => {
+      return engineInfo.binaryPath.right
+    })
+    .with({ binaryPath: P.when(E.isLeft) }, (engineInfo) => {
+      errors.push(engineInfo.binaryPath.left)
+      return 'E_CANNOT_RESOLVE_PATH' as const
+    })
+    .exhaustive()
+
+  const versionMessage = `${version} (at ${path.relative(process.cwd(), absolutePath)}${resolved})`
+  return [versionMessage, errors] as const
+}
+
+export async function resolveEngine(binaryName: BinaryType): Promise<EngineInfo> {
+  const envVar = engineEnvVarMap[binaryName]
+  const pathFromEnv = process.env[envVar]
+
+  const resolvedVersion: { version: E.Either<Error, string> } = await pipe(
+    safeGetBinaryVersion(pathFromEnv, binaryName),
+
+    // "wide" pattern matching, resulting in an union type of the two branches
+    TE.matchW(
+      (versionError) => ({ version: E.left(versionError) }),
+      (version) => ({ version: E.right(version) }),
+    ),
+  )()
+
+  if (pathFromEnv && fs.existsSync(pathFromEnv)) {
+    /**
+     * Extract EngineInfo from a library engine
+     */
+    const engineInfoLibrary: EngineInfoLibrary = {
+      ...resolvedVersion,
+      libraryPath: O.fromNullable(pathFromEnv),
+      fromEnvVar: O.fromNullable(envVar),
+    }
+    return engineInfoLibrary
+  }
```

Comment:
Ok, that was on me because I didn't read this line in `resolveBinary.ts`:

```ts
export const engineEnvVarMap = {
  [BinaryType.queryEngine]: 'PRISMA_QUERY_ENGINE_BINARY',
  [BinaryType.libqueryEngine]: 'PRISMA_QUERY_ENGINE_LIBRARY', // <----
  [BinaryType.migrationEngine]: 'PRISMA_MIGRATION_ENGINE_BINARY',
  [BinaryType.introspectionEngine]: 'PRISMA_INTROSPECTION_ENGINE_BINARY',
  [BinaryType.prismaFmt]: 'PRISMA_FMT_BINARY',
}
```

However, I did test [the most up to date code](https://github.com/prisma/prisma/pull/13737/commits/444b647945ac24b64ae96162026c44c7aed109b7) of this PR, and it still manages env vars for resolving the `library` and the `binary` (as shown in the private chat).
I will rephrase some vars and comments I have introduced in this PR to reflect this fact.

Reply:
In fact, this resulted in a bunch of simplifications 🎉